crypto/sleep.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/crypto.h>
+#include "internal/e_os.h"
+
+/* system-specific variants defining OSSL_sleep() */
+#if defined(OPENSSL_SYS_UNIX) || defined(__DJGPP__)
+
+void OSSL_sleep(uint64_t millis)
+{
+# ifdef OPENSSL_SYS_VXWORKS
+    struct timespec ts;
+
+    ts.tv_sec = (long int) (millis / 1000);
+    ts.tv_nsec = (long int) (millis % 1000) * 1000000ul;
+    nanosleep(&ts, NULL);
+# elif defined(__TANDEM)
+#  if !defined(_REENTRANT)
+#   include <cextdecs.h(PROCESS_DELAY_)>
+
+    /* HPNS does not support usleep for non threaded apps */
+    PROCESS_DELAY_(millis * 1000);
+#  elif defined(_SPT_MODEL_)
+#   include <spthread.h>
+#   include <spt_extensions.h>
+
+    usleep(millis * 1000);
+#  else
+    usleep(millis * 1000);
+#  endif
+# else
+    usleep(millis * 1000);
+# endif
+}
+#elif defined(_WIN32)
+# include <windows.h>
+
+void OSSL_sleep(uint64_t millis)
+{
+    Sleep(millis);
+}
+#else
+/* Fallback to a busy wait */
+# include "internal/time.h"
+
+static void ossl_sleep_secs(uint64_t secs)
+{
+    sleep(secs);
```

Comment:
One problem of course is that `sleep()` is a posix function but not C89 and is defined in unistd.h which may not be universally available? It might be better to use the alternative `select` approach that I proposed because we already compile use of `select` in everywhere  on all platforms except where OPENSSL_NO_SOCK is defined.

Reply:
select is not C89, of course. If there's a platform which has select but not sleep I'd like to know about it. Personally I'm not concerned about this.

Reply:
`select` is also a POSIX function, so we're trading one for another?
Do note that we use `sleep()` in 1.1.1, so it's not like using it hasn't been battle tested already.  Actually, among the things in `e_os.h` is to ensure that we include the correct header file to get `sleep()`, since we used it historically.

Reply:
> Do note that we use sleep() in 1.1.1, so it's not like using it hasn't been battle tested already.

Ah! That's good enough for me.